### PR TITLE
Umount and wait for verity devices to be removed on shutdown

### PR DESCRIPTION
**Pull request by alfonsoros88**

### CODE REVIEW

File: north/src/linux/inotify.rs [hunk 1]
```diff
@@ -0,0 +1,43 @@
+use anyhow::{anyhow, Result};
```
Comment (flxo) [inline north/src/linux/inotify.rs:1]: LIcese header is missing

File: north/src/linux/inotify.rs [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+fn file_does_not_exist(path: &Path) -> bool {
+    match access(path.as_os_str(), AccessFlags::F_OK) {
+        Err(nix::Error::Sys(Errno::ENOENT)) => true,
+        _ => false,
+    }
+}
+
+pub fn wait_for_file_deleted(path: &Path, timeout: Duration) -> Result<()> {
+    let inotify = Inotify::init(InitFlags::IN_CLOEXEC | InitFlags::IN_NONBLOCK)?;
+    inotify.add_watch(path.as_os_str(), AddWatchFlags::IN_DELETE_SELF)?;
```
Comment (flxo) [inline north/src/linux/inotify.rs:20]: This is a blocking operation which is not good in a async context

File: north/src/main.rs [hunk 3]
```diff
@@ -198,6 +199,30 @@ async fn main() -> Result<()> {
 
     info!("Shutting down...");
 
+    // umount all
+    for app in state.applications.values() {
```
Comment (flxo) [inline north/src/main.rs:203]: Check if any app is running before unmounting.

File: north/src/main.rs [hunk 4]
```diff
@@ -198,6 +199,30 @@ async fn main() -> Result<()> {
...
+            .canonicalize()
+            .await
+            .with_context(|| {
+                format!(
+                    "Could not get the verity moint point {}",
+                    app.container().root.display()
+                )
+            })?;
+
+        nix::mount::umount(root_path.as_path().as_os_str())
```
Comment (flxo) [inline north/src/main.rs:217]: use `linux::mount::mount`

File: north/src/linux/inotify.rs [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+use nix::{
+    self,
+    errno::Errno,
+    sys::inotify::{AddWatchFlags, InitFlags, Inotify},
+    unistd::{access, AccessFlags},
+};
+use std::path::Path;
+use std::{os::unix::io::AsRawFd, time::Duration};
+
+fn file_does_not_exist(path: &Path) -> bool {
```
Comment (flxo) [inline north/src/linux/inotify.rs:11]: Remove this function and replace it's call with a `Metadata::is_file` call.

**Review by flxo — CHANGES_REQUESTED** [review]

File: north/src/linux/mount.rs [hunk 6]
```diff
@@ -146,18 +146,9 @@ pub async fn mount(
...
-            0 => Ok(()),
-            _err => Err(anyhow!(
-                "Failed to unmount: {}",
-                std::io::Error::last_os_error()
-            )),
-        }
-    })
-    .await
+    let path = target.to_owned();
+    task::spawn_blocking(move || nix::mount::umount(path.as_os_str()))
```
Comment (flxo) [inline north/src/linux/mount.rs:150]: `mount` uses `libc`. `umount` uses `nix`. This should be the same for both not mix the deps in one mod.

File: north/src/linux/mount.rs [hunk 7]
```diff
@@ -146,18 +146,9 @@ pub async fn mount(
...
+        .await
+        .map_err(anyhow::Error::new)
```
Comment (flxo) [inline north/src/linux/mount.rs:152]: Is this `map_err` needed here? Think the `context` call makes the error a `anyhow::Error` anyway

File: north/src/linux/inotify.rs [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+
+pub async fn wait_for_file_deleted(path: &Path, timeout: Duration) -> Result<()> {
+    let my_path = path.to_owned();
+    task::spawn_blocking(move || {
+        let inotify = Inotify::init(InitFlags::IN_CLOEXEC | InitFlags::IN_NONBLOCK)?;
+        inotify.add_watch(&my_path, AddWatchFlags::IN_DELETE_SELF)?;
+
+        let start_time = std::time::SystemTime::now();
+
+        loop {
```
Comment (flxo) [inline north/src/linux/inotify.rs:34]: Is this spinning until the file is removed? 
`inotify` offers a blocking read that avoids this.

File: north/src/linux/inotify.rs [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+            let elapsed = std::time::SystemTime::now().duration_since(start_time)?;
+            if elapsed > timeout {
```
Comment (flxo) [inline north/src/linux/inotify.rs:36]: This condition would apply nicely in as argument for `while`

File: north/src/linux/inotify.rs [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+pub async fn wait_for_file_deleted(path: &Path, timeout: Duration) -> Result<()> {
+    let my_path = path.to_owned();
+    task::spawn_blocking(move || {
+        let inotify = Inotify::init(InitFlags::IN_CLOEXEC | InitFlags::IN_NONBLOCK)?;
+        inotify.add_watch(&my_path, AddWatchFlags::IN_DELETE_SELF)?;
+
+        let start_time = std::time::SystemTime::now();
+
+        loop {
+            let elapsed = std::time::SystemTime::now().duration_since(start_time)?;
```
Comment (flxo) [inline north/src/linux/inotify.rs:35]: Calculating the time is nicer done with a `Instant`: 

```
let start = std::time::Instant::now();

while start.elapsed() < time::Duration::from_secs(3) {
...
}
```

File: north/src/console.rs [hunk 11]
```diff
@@ -328,7 +328,7 @@ async fn update(state: &mut State, args: &[&str]) -> Result<String> {
 async fn shutdown(state: &mut State) -> Result<String> {
     let stop = stop(state, &[]).await?;
     state.tx().send(Event::Shutdown).await;
-
+    uninstall(state, &[]).await?;
```
Comment (alfonsoros88) [inline north/src/console.rs:331]: move before send

File: north/src/linux/mount.rs [hunk 12]
```diff
@@ -12,152 +12,42 @@
...
-    }
-}
+use std::ffi::CString;
 
 pub async fn mount(
     source: &Path,
     target: &Path,
     fstype: &str,
-    flags: MountFlags,
+    flags: nix::mount::MsFlags,
```
Comment (flxo) [inline north/src/linux/mount.rs:23]: Hm. Using foreign types in public api feels a little bit odd. 
Reexport them with 
```
pub use nix::mount::MsFags as Flags;
```

File: north/src/linux/inotify.rs [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+//   distributed under the License is distributed on an "AS IS" BASIS,
+//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//   See the License for the specific language governing permissions and
+//   limitations under the License.
+
+use anyhow::{Context, Result};
+use async_std::future;
+use async_std::task;
+use nix::sys::inotify::{AddWatchFlags, InitFlags, Inotify};
+use std::fs::metadata;
```
Comment (flxo) [inline north/src/linux/inotify.rs:19]: Isn't `rustfmt` grouping this?
Think a `cargo +nightly fmt --all` is need here.

**Review by flxo — CHANGES_REQUESTED** [review]
Some minor comments. Feel free to merge after changing the mount flag reexport.

File: north/src/linux/inotify.rs (hunk 13 above)
Comment (alfonsoros88) [reply north/src/linux/inotify.rs:19]: something went wrong with my rustfmt

**Pull request merged**